features/mbedtls/targets/TARGET_Silicon_Labs/crypto_sha.c

```diff
@@ -0,0 +1,449 @@
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include "mbedtls/sha1.h"
+#include "mbedtls/sha256.h"
+
+#if defined(MBEDTLS_SHA256_ALT) || defined(MBEDTLS_SHA1_ALT)
+
+#include "em_device.h"
+
+#if defined(CRYPTO_PRESENT)
+#include "em_crypto.h"
+#include "em_core.h"
+#include "crypto_management.h"
+#include "em_assert.h"
+#include <string.h>
+
+static const uint32_t init_state_sha1[8] =
+{
+    0x67452301UL,
+    0xEFCDAB89UL,
+    0x98BADCFEUL,
+    0x10325476UL,
+    0xC3D2E1F0UL,
+    0x0UL,
+    0x0UL,
+    0x0UL
+};
+
+static const uint32_t init_state_sha256[8] =
+{
+    0x6A09E667UL,
+    0xBB67AE85UL,
+    0x3C6EF372UL,
+    0xA54FF53AUL,
+    0x510E527FUL,
+    0x9B05688CUL,
+    0x1F83D9ABUL,
+    0x5BE0CD19UL
+};
+
+static const uint32_t init_state_sha224[8] =
+{
+    0xC1059ED8UL,
+    0x367CD507UL,
+    0x3070DD17UL,
+    0xF70E5939UL,
+    0xFFC00B31UL,
+    0x68581511UL,
+    0x64F98FA7UL,
+    0xBEFA4FA4UL
+};
+
+static const unsigned char sha_padding[64] =
+{
+ 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+};
+
+typedef enum {
+    CRYPTO_SHA1,
+    CRYPTO_SHA2
+} crypto_sha_mode_t;
+
+/*
+ * 32-bit integer manipulation macros (big endian)
+ */
+#ifndef PUT_UINT32_BE
+#define PUT_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (b)[(i)    ] = (unsigned char) ( (n) >> 24 );       \
+    (b)[(i) + 1] = (unsigned char) ( (n) >> 16 );       \
+    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
+    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
+} while( 0 )
+#endif
+
+/**
+ * @brief unified SHA acceleration function.
+ *
+ * @param[inout] state  State context for SHA hashing
+ * @param[in] data      Input block(s). Must contain N blocks
+ *                      of SHA1/SHA256 block size (64 bytes)
+ * @param blocks        Amount of blocks pointed to by data
+ * @param mode          SHA operation mode
+ */
+static void crypto_sha_update_state( uint32_t state[8],
+                                     const unsigned char *data,
+                                     size_t blocks,
+                                     crypto_sha_mode_t mode )
+{
+    CORE_DECLARE_IRQ_STATE;
+    CRYPTO_TypeDef *crypto = crypto_management_acquire();
+
+    switch ( mode ) {
+        case CRYPTO_SHA1:
+            crypto->CTRL = CRYPTO_CTRL_SHA_SHA1;
+            break;
+        case CRYPTO_SHA2:
+            crypto->CTRL = CRYPTO_CTRL_SHA_SHA2;
+            break;
+    }
+
+    crypto->WAC      = 0;
+    crypto->IEN      = 0;
+
+    /* Set result width of MADD32 operation. */
+    CRYPTO_ResultWidthSet(crypto, cryptoResult256Bits);
+
+    /* Clear sequence control registers */
+    crypto->SEQCTRL  = 0;
+    crypto->SEQCTRLB = 0;
+
+    /* Put the state back */
+    CORE_ENTER_CRITICAL();
```

Comment:
Same comment as before on the consecutive reads/writes needed to load a full 128-bit value